### PR TITLE
Fix runtime when gun is destroyed with attached seclight

### PR DESCRIPTION
**Pull request by Timberpoes**
## About The Pull Request

image

The lowdown is that when you have a gun with a seclight attached and throw it in lava, the gun will be destroyed, call QDEL_NULL on any seclight attached.

The attached seclight will be qdel'd. This will then call handle_atom_del() on the gun, which will call clear_gunlight() because the seclight hasn't been nulled on the gun yet, it's still in the middle of its own qdel.

clear_gunlight doesn't early return as gun_light on the gun itself isn't null yet, so set_gun_light(null) gets called. This eventually leads to the seclight being forceMove()ed to the turf the gun is on in the middle of the seclight's qdel. If that turf is, for instance, lava - The light will immediately start taking fire damage and thus the runtime happens.

I put a check in to make sure the forceMove logic is skipped if the seclight is being QDELETED. In testing, this was enough to guard against the runtime.

Basically - QDEL_NULL does what it says on the tin. qdels, then nulls after the qdel. Because of this, there's an oversight that allows a seclight to be forceMove()ed off the gun in the middle of the seclight's qdel and thus take damage while being qdel'd.

## Why It's Good For The Game

Feex runtimes.

### CODE REVIEW

**Comment by Rohesie** [comment]
Might be better to do something like this:
```nix
var/backup = var_to_del
var_to_del = null
qdel(backup)
```

**Comment by Timberpoes** [comment]
> 
> 
> Might be better to do something like this:
> 
> ```c++
> var/backup = var_to_del
> var_to_del = null
> qdel(backup)
> ```

Oh great. I wrote a reply to this and clearly didn't submit it.

The null-then-qdel still allows a runtime to happen if the seclight is qdel'd from the gun via some other source than the gun itself being Destroy()ed.

But that proc is a nightmare. I've cleaned it up a bit, let me know your thoughts.

**Comment by Rohesie** [comment]
Looks good, but preserve the `. = gun_light` behavior so that proc redefinitions can know after calling parent if there was a change or not.

**Comment by Timberpoes** [comment]
> 
> 
> Looks good, but preserve the `. = gun_light` behavior so that proc redefinitions can know after calling parent if there was a change or not.

Done with additional code docs about the return value.
